experimental/algorithm/LAGr_Modularity.c

```diff
@@ -1,61 +1,110 @@
+
+// LAGraph, (c) 2019-2022 by The LAGraph Contributors, All Rights Reserved.
+// SPDX-License-Identifier: BSD-2-Clause
+//
+// For additional details (including references to third party source code and
+// other files) see the LICENSE file or contact permission@sei.cmu.edu. See
+// Contributors.txt for a full list of contributors. Created, in part, with
+// funding and support from the U.S. Government (see Acknowledgments.txt file).
+// DM22-0790
+
+// Contributed by Cameron Quilici, Texas A&M University
+
+//------------------------------------------------------------------------------
+
+// The modularity (Q) of a graph clustering C is defined as (directed case):
+//
+// Q = \sum{c = 1}^n [\frac{L_c}{m} - \gamma(\frac{k_c^{in} k_c^{out}}{m})^2]
+//
+// where the sum iterates over all communities c = 1,...,n, L_c is the the
+// number of total edges in cluster c, m is the total number of edges in the
+// graph, k_c^{in} and k_c^{out} are the total in and out degrees of cluster c,
+// and \gamma is the resolution parameter which controls the relative importance
+// of the intra-cluster and inter-cluster edges.
+
+// Modularity works by comparing the intra-cluster density of a particular
+// clustering to that of a random graph with the same degree distribution. The
+// range of Q is [-.5, 1]. When Q ~ 0, this indicates that the clustering is
+// not much better than at random whereas Q ~ 1 indicates a strong community
+// structure.
+
+// https://arxiv.org/abs/0906.0612 pp. 15-16
+
+#define LG_FREE_WORK                                                           \
+    {                                                                          \
+        GrB_free(&l);                                                          \
+        GrB_free(&vmask);                                                      \
+        GrB_free(&k_in);                                                       \
+        GrB_free(&k_out);                                                      \
+        GrB_free(&in_degree);                                                  \
+        GrB_free(&out_degree);                                                 \
+        GrB_free(&CA);                                                         \
+        GrB_free(&C);                                                          \
+        GrB_free(&ONE_BOOL);                                                   \
+        LAGraph_Free((void **)&lX, NULL);                                      \
+        LAGraph_Free((void **)&k_outX, NULL);                                  \
+        LAGraph_Free((void **)&k_inX, NULL);                                   \
     }
 
-#define LG_FREE_ALL   \
-    {                 \
-        LG_FREE_WORK; \
+#define LG_FREE_ALL                                                            \
+    {                                                                          \
+        LG_FREE_WORK;                                                          \
```

Comment:
not freeing the double pointer handles, not sure if I need to or if that's just for GB objects